### PR TITLE
Tokenizer implementation

### PR DESCRIPTION
**Pull request by kiyomih**
**Implementation for Tokenizer.**
- Updated Tokenizer so we can initialize multiple if we want to (easier for testing, consistent with Arthur's TS example)
- exposes constructor, getNext, checkToken, getAndCheckNext
- each new line starts with token of format: _INDENT_LEVEL=[0-9]+_
- tokenize: tokenizes program into string[][]. Does not tokenize anything inside quotation marks.

**Other updates:**
- Updated ASTNodes to pass Tokenizer context
- Added readFile for FileSystem wrapper

**TODO:**
- Need to also have a better way to handle '[' and ']' chars for tokenize (in the case user wants to specify an array type.
- Feel free to add more examples/ tests for Tokenize- there's a lot of different cases that need to be covered.

### CODE REVIEW

File: test/util/Tokenizer.spec.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+describe("Tokenizer tokenize", () => {
+    it("gets and checks tokens in projectstructureex", () => {
+        let tokenizer: Tokenizer = new Tokenizer("projectstructureex.txt", "./test/testFiles");
+        //dir src
+        tokenizer.getNext();
+        expect(tokenizer.getNext()).to.equal("dir");
+        expect(tokenizer.getNext()).to.equal("src");
+        // 	class Time
+        tokenizer.getNext();
```
Comment (scveloso) [inline test/util/Tokenizer.spec.ts]: this should be an INDENT_LEVEL = 8 right? 

dont have to change it, just checking my understanding of how you're dealing with tabs

**Review by scveloso — COMMENTED** [review]
Thanks for working on this! Trying to wrap my head around how we can parse tabs.

So if we are one dir in (in the parsing stage), should we keep doing getAndCheckNext() and if we encounter INDENT_LEVELs that don't match our current level, then we have left this current level's scope and we finish parsing this dir?

File: test/util/Tokenizer.spec.ts (hunk 1 above)
Comment (kiyomih) [reply test/util/Tokenizer.spec.ts]: yep- should be `_INDENT_LEVEL=8_`
One `\t` will increase the space count by 8

**Comment by kiyomih** [comment]
@scveloso - thanks for reviewing!

Yep- I was thinking if we encounter something that is of an indentation level < than the dir's indentation level, that would mean that we are outside of that directory.

At a high-level, something like:
```
parse(context: Tokenizer): any {
    let indentLevel = this.extractTabLevel(context.getNext());
    context.getAndCheckNext("dir");
    while(this.extractTabLevel(!context.checkToken(Tokenizer.NO_MORE_TOKENS) && this.extractTabLevel(context.getNext()) > indentLevel) {
        // calling parse for Content type, adding to Content[]
     }
}
```

*Note that checkToken() will not actually move the position we are in the tokens array, but calls to getNext() will.

**Review by jyoo980 — APPROVED** [review]
@kiyomih @scveloso 

LGTM! The only question I had was Sam's but that looks like it was addressed!

**Comment by kiyomih** [comment]
Updated slightly so we abstract away the indentation calculation all together from the AST nodes, after realizing that it might be easier to implement parse if we do that.
See the ClassDecl for an example on how to use latest iteration of Tokenizer.*

*This implementation hasn't been tested since this would require children node parse() functions to have been implemented, this is just an example of how we could interact with the Tokenizer in parse

Changes made: 
- removed extractTabLevel from ASTNode
- add getCurrentLineTabLevel to Tokenizer
- getNext skips over the INDENT_LEVEL tokens (parse caller never sees them)

What do y'all think? Is this easier to use that previous version? (If previous version is preferred can just revert, so its easy to change back)

**Comment by jyoo980** [comment]
LGTM, but I'll wait for @scveloso to have a look over as well!

File: src/util/Tokenizer.ts [hunk 2]
```diff
@@ -1,60 +1,225 @@
...
+    public static NO_MORE_TOKENS = "NO_MORE_TOKENS";
+    private static reservedTokenWord = "_RESERVED_";
+
 
-    private constructor(fileName: string) {
-        Tokenizer.tokenizerFile =  fileName;
-        // TODO: implement this
+    constructor(fileName: string, path: string) {
+        // initialize fields
+        this.tokenizerFile =  `${path}/${fileName}`;
```
Comment (daviidli) [inline src/util/Tokenizer.ts]: not needed?

**Review by daviidli — APPROVED** [review]
Looks good!

**Review by scveloso — APPROVED** [review]
lgtm! thanks for working on this

File: src/util/Tokenizer.ts (hunk 2 above)
Comment (kiyomih) [reply src/util/Tokenizer.ts]: Thanks! removed this in latest commit (only change made in commit)

**Pull request merged**